src/context.ts

```diff
@@ -53,14 +52,22 @@ export class OperonContextImpl implements OperonContext {
   /*** Application configuration ***/
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   applicationConfig?: any;
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  getConfig(key: string): any {
-    if (!this.applicationConfig) {
-      return undefined;
+  getConfig<T>(key: string, defaultValue?: T): T {
+    // If there is no application config at all, or the key is missing, return the default value or undefined.
+    if (!this.applicationConfig || !has(this.applicationConfig, key)) {
+      if (defaultValue) {
+        return defaultValue;
+      }
+      return undefined as T;
```

Comment:
This seems odd rather than having `getConfig` return `T | undefined`.

I'm thinking we [overload](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads) the getConfig function like this:

```ts
getConfig<T>(key: string): T | undefined;
getConfig<T>(key: string, defaultValue: T): T;
```

Reply:
> This seems odd rather than having getConfig return T | undefined.

I agree. With that change, we don't need to overload, I trust?

Reply:
I think you do. If you provide a default, then you can never get undefined. If you don't provide a default, then you have to specify the type arg explicitly *and* it's possible that you get undefined.

Reply:
I added some tests where I don't provide the type explicitly, for both a missing key and an present one.

```typescript
      // Config key exists
      expect(ctx.getConfig("payments_url")).toBe("http://somedomain.com/payment");
      // Config key does not exist, no default value
      expect(ctx.getConfig("no_key")).toBeUndefined();
      // Config key does not exist, default value
      expect(ctx.getConfig("no_key", "default")).toBe("default");
```

And they seem to function. I'm not sure what we are getting from the overload?

Reply:
you're testing runtime behavior not compile time/intellisense. I think you just need to provide the two overload declarations and then change the implementation to return `T | undefined`.